### PR TITLE
avm1: implement hscroll and maxhscroll for TextField

### PR DESCRIPTION
**Pull request by midgleyc**
Implement hscroll and maxhscroll for avm1's TextField.

avm2 looks similar with `scrollH` / `maxScrollH` properties, but I don't have a way of generating avm2 files to test with.

### CODE REVIEW

**Comment by midgleyc** [comment]
The current test for hscroll uses very small boxes, with an epsilon for "close". I think this is in part a font thing: any decently large text falls out of sync quickly. In the following image, Flash gets an extra "e" in.
image

Also, my Flash MX default font seems to be Times New Roman, while Ruffle renders it as Arial.

Some of the other edittext tests don't seem to have this problem: me using an old IDE might be exacerbating it.

File: core/src/avm1/globals/text_field.rs [hunk 1]
```diff
@@ -589,3 +591,28 @@ pub fn set_type<'gc>(
...
+
+pub fn set_hscroll<'gc>(
+    this: EditText<'gc>,
+    activation: &mut Activation<'_, 'gc, '_>,
+    value: Value<'gc>,
+) -> Result<(), Error<'gc>> {
+    let hscroll_pixels = value.coerce_to_f64(activation)?;
+    let clamped = hscroll_pixels.floor().clamp(0.0, this.maxhscroll());
+    this.set_hscroll(clamped, &mut activation.context);
+    Ok(())
```
Comment (Herschel) [inline core/src/avm1/globals/text_field.rs:610]: From my testing, seems like this has some weird absolute value behavior, as well as `i32` wrapping in twips, so this should probably be `coerce_to_i32` and `hscroll` possibly stored directly in Twips. Haven't sussed it out completely, but:

```
function testHscroll(n) {
	txt.hscroll = n;
	trace("// " + n + ": " + txt.hscroll);
}

testHscroll(-10); // 10: absolute value 
testHscroll(107374179); // txt.maxhscroll, clamped
testHscroll(107374180); // 0? (near 2^31 twips, possibly affected by 4px padding margin?)
testHscroll(107374181); // 0
testHscroll(107374182); // 0
testHscroll(107374183); // 0
testHscroll(107374184); // 0
testHscroll(107374185); // 0
testHscroll(107374186); // txt.maxhscroll -- we've wrapped around to ~-2^31 twips, the absolute value kicks in again)
testHscroll(214748364 - txt.maxhscroll); // txt.maxhscroll
testHscroll(214748364 - txt.maxhscroll + 1); // txt.maxhscroll - 1, approaching 0
testHscroll(214748363); // 1
testHscroll(214748364); // 0, near 2^32 twips
testHscroll(214748365); // 0
testHscroll(214748366); // back positive, 1
testHscroll(214748367); // 2
testHscroll(214748368); //3
```

Comment (Herschel) [reply core/src/avm1/globals/text_field.rs:610]: Yikes, looked into this a bit more, and this is a version difference between SWFv8 and SWFv9.

SWFv8 and earlier match your PR (using the above test):
```
// -10: 0
// 107374179: 5344
// 107374180: 5344
// 107374181: 5344
// 107374182: 5344
// 107374183: 5344
// 107374184: 5344
// 107374185: 5344
// 107374186: 5344
// 214743020: 5344
// 214743021: 5344
// 214748363: 5344
// 214748364: 5344
// 214748365: 5344
// 214748366: 5344
// 214748367: 5344
// 214748368: 5344
```

**Comment by midgleyc** [comment]
There's also some odd behaviour with embedded font characters the leads to drastically increasing the `maxhscroll` and lowering the amount scrolled. I expect it's a bug, and hope it goes away at higher versions. No attempt has been made to replicate this behaviour.

embedded-chars-hscroll

**Review by Herschel — APPROVED** [review]

**Comment by Herschel** [comment]
Thanks!

**Pull request merged**